pkg/controller/job/job_controller.go

```diff
@@ -968,8 +968,8 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 			uidsWithFinalizer.Insert(uid)
 		}
 	}
-	// Shallow copy, as it will only be used to detect changes in the counters.
-	oldCounters := job.Status
+
+	oldJob := job.DeepCopy()
```

Comment:
`DeepCopy` is generally expensive. Why is a shallow copy not enough?

Reply:
A shallow copy results in most of the tests failing due to this line I believe: https://github.com/kubernetes/kubernetes/pull/115349/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R1103 With a shallow copy of `job` the `oldJob.Status` pointer is still pointing to the same address in memory as the original `job` so it is modifying the wrong status.

Reply:
Refactored succeededIndexesFromJob to accept a `batch.JobStatus` and `completions` and calculate the succeeded metric diff for Indexed Jobs using the *new job spec's completions* like so, which seems to work:

```go
succeededIndexesFromJob(job.Status.CompletedIndexes, int(*job.Spec.Completions)).total() - succeededIndexesFromJob(oldCounters.CompletedIndexes, int(*job.Spec.Completions)).total()
```